Fix carousel sliding direction in RTL pages

aqeelat opened the pull request:
Both LTR and RTL carousels slid in the same direction.

Fixed and tested.

aqeelat committed:
fixed movement direction to reflect off of the ltr version.

ffoodd commented:
Could you provide a reduced test case please, or describe exactly what's wrong?

Because as of v5.0.0-beta2, they don't slide in the same direction:
- [RTL cheatsheet](https://getbootstrap.com/docs/5.0/examples/cheatsheet-rtl/#carousel)
- [LTR cheatsheet](https://getbootstrap.com/docs/5.0/examples/cheatsheet/#carousel)

Am I missing something?

AqeelAT commented:
When the slides are cycling, they both move in the same direction, which should not be the case.
This is what it looks like after the fix:
![sliding direction](https://imgur.com/qEkkI4O.gif)

Before the fix, the RTL version would slide in the same way as the LTR.

XhmikosR commented:
We need tests, please add some tests.

AqeelAT commented:
I'm working on the tests but I'm having trouble coming up with a way to mock index.isRTL. Any suggestions?

aqeelat committed:
1- Added unit tests
2- made it so that rtl carousels can be used in ltl pages, by using a localized dir='rtl' tag. This could be replaced by a designated class.

aqeelat committed:
Merge branch 'main' of https://github.com/twbs/bootstrap into rtl-carousel

aqeelat committed:
Merge branch 'main' into rtl-carousel

AqeelAT commented:
Updated the pull request with unit tests.
@XhmikosR @ffoodd

aqeelat committed:
deleted commented code and fixed typo

aqeelat committed:
Merge remote-tracking branch 'origin/rtl-carousel' into rtl-carousel

ffoodd commented:
I'd stick with `isRTL` instead of  adding it to the constructor—and for the least if you want to add it, this should be done in other components using it for consistency.

But JS is not my specialty. @rohit2sharma95 any opinion on this?

XhmikosR commented on js/src/carousel.js:
```diff
@@ -449,9 +450,20 @@ class Carousel extends BaseComponent {
+    // | F     | T                            | Option 1 |
+    // | F     | F                            | Option 2 |
+    // +-------+------------------------------+----------+
+    const directionalClassName = this.isRTL === (direction === DIRECTION_NEXT) ? CLASS_NAME_END : CLASS_NAME_START
```
Not sure I like these, I feel like it's less readable...

XhmikosR commented on js/src/carousel.js:
```diff
@@ -120,6 +120,7 @@ class Carousel extends BaseComponent {
     this._touchSupported = 'ontouchstart' in document.documentElement || navigator.maxTouchPoints > 0
     this._pointerEvent = Boolean(window.PointerEvent)
 
+    this.isRTL = this._element.dir === 'rtl' || isRTL
```
Why is the new check needed? Or why doesn't the previous one suffice?

BTW, I have a draft PR around to switch `isRTL` to a function so that it's tree-shaken, but I doubt it's related.

AqeelAT replied:
I had to add it so that I can produce the unit tests I was asked for. Then I realized it might be helpful in some cases.

isRTL only checks if the <HTML> tag has dir=rtl attribute. This means that you can't have an RTL carousel otherwise. One use case: in the documentation for bootstrap itself on how to make a page rtl. The page itself could be in English, but the carousel could be in Arabic.

XhmikosR replied:
I don't think we support such a case in v5 @ffoodd ?

rohit2sharma95 replied:
Yes, that case is not supported yet. And this check is then redundant :slightly_smiling_face:

XhmikosR replied:
@AqeelAT: I merged #32446. Your use case makes sense, but I thought it wasn't supported.

That being said, it seems it's sort of supported https://getbootstrap.com/docs/5.0/getting-started/rtl/#ltr-and-rtl-at-the-same-time, so after #32913 or in  #32913, we could refactor the isRTL function to check for an element's direction and or `document.documentElement.dir`.

We just need to make sure we have JS tests to cover these changes

XhmikosR commented:
We also have #32913 FYI

XhmikosR commented:
@rohit2sharma95 @ffoodd I think that #32913 is better to understand and it should have the same fix. Please have a look there and let's close this one.

@AqeelAT thanks for the PR, regardless :)

XhmikosR commented:
Closing in favor of #32913. We can revisit later if needed.

rohit2sharma95 commented:
> I'd stick with `isRTL` instead of adding it to the constructor

I am personally in favor of #32446 so that dir can be checked in run time. @ffoodd
